src/glib/glib.c

```diff
@@ -0,0 +1,491 @@
+        abort();
+    }
+
+    nats_initForOS();
+
+    // Setup a hook for when the process exits.
+    atexit(natsLib_Destructor);
+}
+
+
+static void
+_libTearDown(void)
+{
+    nats_waitForDispatcherPoolShutdown(&gLib.messageDispatchers);
+    nats_waitForDispatcherPoolShutdown(&gLib.replyDispatchers);
+
+    if (gLib.timers.thread != NULL)
+        natsThread_Join(gLib.timers.thread);
+
+    if (gLib.asyncCbs.thread != NULL)
+        natsThread_Join(gLib.asyncCbs.thread);
+
+    if (gLib.gc.thread != NULL)
+        natsThread_Join(gLib.gc.thread);
+
+    natsLib_Release();
+}
+
+// environment variables will override the default options.
+natsStatus
+nats_openLib(natsClientConfig *config)
+{
+    natsStatus s = NATS_OK;
+
+    natsClientConfig defaultConfig = {
+        .LockSpinCount = -1,
+        .ThreadPoolMax = 1,
+    };
+    if (config == NULL)
+        config = &defaultConfig;
+
+    if (!nats_InitOnce(&gInitOnce, _doInitOnce))
+        return NATS_FAILED_TO_INITIALIZE;
+
+    natsMutex_Lock(gLib.lock);
+
+    if (gLib.closed || gLib.initialized || gLib.initializing)
+    {
+        if (gLib.closed)
+            s = NATS_FAILED_TO_INITIALIZE;
+        else if (gLib.initializing)
+            s = NATS_ILLEGAL_STATE;
+
+        natsMutex_Unlock(gLib.lock);
+        return s;
+    }
+
+    gLib.initializing = true;
+    gLib.initAborted = false;
+
+#if !defined(_WIN32)
+    signal(SIGPIPE, SIG_IGN);
+#endif
+
+    srand((unsigned int)nats_NowInNanoSeconds());
+
+    gLib.refs = 1;
+
+    // If the caller specifies negative value, then we use the default
+    if (config->LockSpinCount >= 0)
+        gLockSpinCount = config->LockSpinCount;
+
+    gLib.config = *config;
+    nats_Base32_Init();
+
+    s = natsCondition_Create(&(gLib.cond));
+
+    if (s == NATS_OK)
+        s = natsCrypto_Init();
+
+    if (s == NATS_OK)
+        s = natsMutex_Create(&(gLib.timers.lock));
+    if (s == NATS_OK)
+        s = natsCondition_Create(&(gLib.timers.cond));
+    if (s == NATS_OK)
+    {
+        s = natsThread_Create(&(gLib.timers.thread), nats_timerThreadf, &gLib);
+        if (s == NATS_OK)
+            gLib.refs++;
+    }
+
+    if (s == NATS_OK)
+        s = natsMutex_Create(&(gLib.asyncCbs.lock));
+    if (s == NATS_OK)
+        s = natsCondition_Create(&(gLib.asyncCbs.cond));
+    if (s == NATS_OK)
+    {
+        s = natsThread_Create(&(gLib.asyncCbs.thread), nats_asyncCbsThreadf, &gLib);
+        if (s == NATS_OK)
+            gLib.refs++;
+    }
+    if (s == NATS_OK)
+        s = natsMutex_Create(&(gLib.gc.lock));
+    if (s == NATS_OK)
+        s = natsCondition_Create(&(gLib.gc.cond));
+    if (s == NATS_OK)
+    {
+        s = natsThread_Create(&(gLib.gc.thread), nats_garbageCollectorThreadf, &gLib);
+        if (s == NATS_OK)
+            gLib.refs++;
+    }
+    if (s == NATS_OK)
+        s = natsNUID_init();
+
+    if (s == NATS_OK)
+        s = nats_initDispatcherPool(&(gLib.messageDispatchers), config->ThreadPoolMax);
```

Comment:
It just initializes the struct (mutex), and if `config->ThreadPoolMax` is 0 makes no other allocations. No threads are started until needed. It sort of fell out of using a pre-allocated struct for messageDispatchers in gLib; since it's always there and no easy way to check `== NULL` if it has been initialized, best to have it pre-initialized?

Reply:
Ok.